### ISSUE TITLE
float32: five linear 4*pi*d_L^2 sites are inf at every distance, and the log-domain site loses its distance gradient

### ISSUE THREAD

**Issue by suchethac** [suchethac/tengri#1859]
Sibling of #1837 at the opposite end of the exponent range, and a second, harder defect found while measuring it. Child of #1206; directly relevant to #1388.

Tier A moved the cosmological flux scale to log-domain at **one** site. Five JAX-traced siblings still materialize `4*pi*d_L**2` linearly, and every one is `inf` in float32 at every astronomical distance. Separately, the *correct* log-domain site loses its distance gradient in float32 — value right, derivative wrong-signed.

## Part 1 — five linear sites, `inf` at every distance

The correct form, and the only site that has it:

```python
# src/tengri/observation/redshift_kernel.py:92-93
log10_flux_scale = jnp.log10(1.0 + z) - LOG10_4PI - 2.0 * jnp.log10(dl_cm)
return apply_log10_scale(L_on_obs_grid, log10_flux_scale)
```

The five that do not:

| site | expression |
|---|---|
| `utils/grid_interp.py:498` | `(1.0 + redshift) / (4.0 * _math.pi * dl_cm**2)` |
| `components/stellar/sps/precompute.py:460` | `(1.0 + redshift) / (4.0 * _math.pi * dl_cm**2)` |
| `forward/sed_model.py:5458` | `four_pi_dl2 = 4.0 * jnp.pi * dl_cm**2` |
| `measure.py:158` (`_four_pi_dl2`) | `4.0 * jnp.pi * jnp.asarray(dl_cm) ** 2` |
| `components/nebular/line_precompute.py:91` (`_four_pi_dl2`) | `4.0 * jnp.pi * dl_cm**2` |

The first two carry comments stating explicitly that `dl_cm` and `redshift` arrive as **tracers**, so they are on the traced path, not build-time. `precompute.py:460` is on the `WavePrecomp` path — the default for every fit (`approx="auto"`).

`components/stellar/sps/precompute.py:960` uses `np.pi` eagerly in float64 on the build path and is **not** affected.

Measured, against float32 max `3.4028235e38`:

| distance | `d_L` [cm] | `d_L**2` f64 | f32 |
|---|---|---|---|
| 10 pc | 3.0857e+19 | 9.5214e+38 | `inf` |
| z=0.01 | 1.3268e+26 | 1.7605e+52 | `inf` |
| z=0.5 | 8.9485e+27 | 8.0075e+55 | `inf` |
| z=3 | 8.0228e+28 | 6.4365e+57 | `inf` |

It overflows **at the square**, at every distance — there is no safe redshift, including 10 pc. Exactly the shape of #1837 (`TEN_PC_CM**2 = 9.52e38`, input 2.4e27 fine, output −16.8 mag fine, `inf` in between).

Consequence, measured on the projected flux with `L_nu = 1e28 erg/s/Hz`:

```
=== f32 ===
  z=0.01   linear: F_nu=0.000000e+00  d/dz=0.000000e+00     log-applied: F_nu=4.234632e-26  d/dz=+4.192706e-26
  z=0.5    linear: F_nu=0.000000e+00  d/dz=0.000000e+00     log-applied: F_nu=1.469142e-29  d/dz=+9.794278e-30
  z=3.0    linear: F_nu=0.000000e+00  d/dz=0.000000e+00     log-applied: F_nu=4.923954e-31  d/dz=+1.230989e-31
=== f64 (both forms identical) ===
  z=0.01   F_nu=4.234615e-26  d/dz=-8.491363e-24
  z=0.5    F_nu=1.469156e-29  d/dz=-6.054209e-29
  z=3.0    F_nu=4.923993e-31  d/dz=-2.710173e-31
```

`(1+z)/inf` is exactly `0.0`, so the failure is **silent** — the flux is not NaN, it is zero, and a finiteness check passes.

## Part 2 — the log-domain site loses its distance gradient

The `log-applied` column above gets the **value** right (4.234632e-26 vs 4.234615e-26, rel 4e-6) and the **gradient** wrong-signed at all three redshifts.

The f32 gradient / value ratios are `0.9901`, `0.667`, `0.2500` — exactly `1/(1+z)`. So this is not a sign error but a **term annihilation**: of

```
s(z) = log10(1+z) - LOG10_4PI - 2*log10(d_L(z))
```

the `-2*log10(d_L)` term contributes exactly zero, leaving `F/(1+z)`, which is positive while the true derivative is negative.

**Mechanism, confirmed by a controlled sweep.** The reverse pass forms the cotangent on `d_L` as `~2F/d_L` — a product of two *small* numbers. Sweeping `d_L` at fixed `L_nu = 1e28`, parametrised as `d_L = d_L0*exp(u)` so `d/du` is well-scaled throughout:

```
  dl_cm     F_nu(f64)     |cot_dl|     d/du f64        d/du f32
  1e+19   7.9577e-12   1.592e-30   -1.5915e-11   -1.5916e-11   OK
  1e+20   7.9577e-14   1.592e-33   -1.5915e-13   -1.5916e-13   OK
  1e+21   7.9577e-16   1.592e-36   -1.5915e-15   -1.5916e-15   OK
  1e+22   7.9577e-18   1.592e-39   -1.5915e-17   -0.0000e+00   ANNIHILATED
  1e+23   7.9577e-20   1.592e-42   -1.5915e-19   -0.0000e+00   ANNIHILATED
  1e+24   7.9577e-22   1.592e-45   -1.5915e-21   -0.0000e+00   ANNIHILATED
  1e+25   7.9577e-24   1.592e-48   -1.5915e-23   -0.0000e+00   ANNIHILATED
  ...
  1e+28   7.9577e-30   1.592e-57   -1.5915e-29   -0.0000e+00   ANNIHILATED
```

The crossover sits between `1.592e-36` and `1.592e-39`, i.e. at float32's smallest **normal** (`1.175e-38`) — not the smallest subnormal (`1.401e-45`), so subnormals are being flushed to zero. In isolation the term shows its true form, exactly `-0.0`; it only *looks* like a sign flip when the surviving `(1+z)` term carries the sum.

Every value and every derivative in this chain is comfortably representable. Only the intermediate cotangent is not. This is the mirror of #1837 — same "fine at both ends, broken in the middle", opposite end of the exponent range.

**It cannot be fixed by rescaling the flux.** Lifting `2F/d_L` above `1.175e-38` at `d_L ~ 8.9e27` needs `F > 5.2e-11`, i.e. `L_nu > 4.2e38` — above float32's max of `3.4e38`. The input overflows before the cotangent becomes representable. The *derivative* has to be taken in log space, not merely the value.

## Why an audit misses both

`d_L` (8.9e27) is representable; `F_nu` (1.5e-29) is representable; `log10_flux_scale` (−56.8) is representable. Both defects live entirely in intermediates — a squared constant in Part 1, a product of two small numbers in Part 2. Sampling inputs and outputs scores the chain safe. Instrument the *chain*, and for Part 2 the forward pass alone is not enough: the value is correct to 4e-6 while the gradient points the wrong way. A wrong-signed gradient is worse than a NaN operationally — NaN halts the optimizer, a sign flip walks it confidently uphill, and no finiteness check can see it.

## Suggested fix

- **Part 1:** the five sites take the `redshift_kernel.py:92-93` form — never materialize the scale, hand the log offset to `apply_log10_scale`. Note this is only correct if the scale is *applied*; `10**(log10(1+z) - LOG10_4PI - 2*log10(dl))` is a linear form wearing a log hat and is `0.0` in f32 just like the original (measured — it caught me while writing this).
- **Part 2:** carry the distance contribution as `log10(d_L)` end-to-end so the derivative is taken with respect to the log, rather than reconstructing `d_L` and differentiating through it. This is #1388's territory and is the concrete argument for it.
- **Guard:** an AST two-way gate banning `4*pi*d_L**2` in traced code, in the shape of `tests/regression/precision/test_no_raw_erg_s_derived_read.py` (#1847). A one-site fix beside five untouched siblings is what produced this.

## Provenance

Part 2 was independently observed by the session working #1837/#1847 (`dl -> interp -> log10_flux_scale -> apply_log10_scale`: f64 `-2.905708e-26`, f32 `+5.607974e-27`, "finite but sign-flipped and ~5x off"). The reproduction above is minimal — ~40 lines, no SSP grid, no model build — and the sweep locates the boundary.

Probes: `probe_fluxscale2.py`, `probe_cotangent2.py`.
Measured on `7abb077bd`, CPU, `jax.config.update` after `import tengri` (#1840).

**Comment by suchethac** [comment]
## Threshold tightened by a factor of ~100, by combining two sweeps that disagree

Independently reproduced by the session working #1837/#1847, rebuilt from the primitives rather than from my probe. Both findings confirmed:

**The `1/(1+z)` signature**, matching to every displayed digit:

| z | f64 `(d/dz)/F` | f32 `(d/dz)/F` | `1/(1+z)` |
|---|---|---|---|
| 0.01 | −200.522679 | 0.990099 | 0.990099 |
| 0.50 | −4.120876 | 0.666667 | 0.666667 |
| 3.00 | −0.550401 | 0.250000 | 0.250000 |

The f64 ratio is nowhere near `1/(1+z)` and is opposite in sign, so the f32 agreement is not a coincidence of scale — it is the `-2*log10(d_L)` addend contributing exactly zero and the `log10(1+z)` addend carrying the sum alone.

**The threshold.** The two sweeps report different `|cot|` at the same distance, and the difference is exactly `4*pi` (12.5664 measured, to six digits) — their chain normalises `F = L/d_L**2`, mine `F = L/(4*pi*d_L**2)`. That makes the two sweeps sample the boundary at *offset* points, which is what tightens it:

| sweep | `d_L` | \|cot\| ~ `2F/d_L` | f32 `d/du` | |
|---|---|---|---|---|
| theirs | 1e+22 | 2.000e-38 | −2.0000e-16 | **survives** |
| mine | 1e+22 | 1.592e-39 | −0.0000e+00 | annihilated |
| theirs | 1e+23 | 2.000e-41 | −0.0000e+00 | annihilated |

Combined bracket: **survives at 2.000e-38, dies at 1.592e-39** — a factor of 12.6, straddling

```
float32 finfo.tiny (smallest normal) = 1.175494e-38   <- the boundary
float32 smallest_subnormal           = 1.401298e-45
```

Neither sweep alone is tight (each spans ~1000x). The *union* pins it to 12.6x precisely **because** the two normalisations differ — two identical probes would have bought nothing. Subnormals are flushed to zero in traced code; the usable float32 floor is `1.175494e-38`, seven decades above the smallest subnormal.

**Corollary, and it is reassuring.** `utils/scale.py::representable_floor` already resolves to `finfo(dtype).tiny`, i.e. the smallest **normal** — so the helper is correct under this finding and the #1847 migrations onto it are sound. The exposure is hand-rolled range reasoning written against `smallest_subnormal`, which is the #1492 inert-guard-floor class.

The AST census guard for Part 1 will be a follow-up PR extending `tests/regression/precision/test_no_raw_erg_s_derived_read.py` with a second banned pattern, keeping one two-way guard rather than two.

**Comment by suchethac** [comment]
## Part 1 fixed in #1878 — and two corrections to this issue as filed

Both corrections came from measuring rather than reading, and both matter to anyone acting on the text above.

**1. `sps/precompute.py:960` is affected. I cleared it here, and I was wrong.**

The issue says it "uses `np.pi` eagerly in float64 on the build path and is **not** affected". That is true of the *build* and irrelevant to the *contract*: `jnp.array(flux_scale_all)` then stores the result, and the cast zeroes every entry.

```
z=0.00   f64 8.3577e-41   f32 8.3578e-41   (a subnormal, survives the cast)
z=0.01   f64 4.2346e-54   f32 0.0000e+00
z=0.50   f64 1.4692e-57   f32 0.0000e+00
z=3.00   f64 4.9240e-59   f32 0.0000e+00
```

Reading the code answered a question I had not asked. A correct float64 build is not a safe float32 *storage*, and for a table the storage is the contract.

**2. `l_line` overflows independently of distance, so the divisor was only half the defect.**

`(feat - cont) * (c/λ_c²) * Δλ` is 1.3913e+40 erg/s for an ordinary galaxy at *every* redshift, against a float32 max of 3.4028e+38. So even with a perfect `4πd_L²`, the line path is `inf/inf = nan`. `_line_flux_from_means` now folds the conversion constants *and* the distance into a single log offset applied to the O(1e28) window mean, so neither end is materialized.

**3. Not a correction, but context I did not have when filing:** these five sites were never undiscovered. They were knowingly deferred under #1186 as "Task 6", and `tests/regression/precision/test_no_raw_flux_scale.py` already listed each one with a rationale. What was missing was not the inventory but the *consequence* — the allow-list recorded "needs log-table threading", never "returns `nan`". That allow-list is now empty.

Relatedly, `docs/dev/float32-tier-b-boundary.md` called this residue "latent rather than live", inferred from a pure-float32 photometry prediction being finite at 8.3e-6 relative — which it is. That inference was sound about photometry and silent about lines. The doc now records the correction rather than dropping it.

## Scope of #1878

- **Line path: live fix.** `measure_line_fluxes`, `predict_line_fluxes`, `measure.line_flux` were `nan` in float32 at every redshift.
- **Photometry path: preventive.** Those stored `flux_scale` scalars have no reader in `src/` today — the live seams were converted under Tier A — so that half converts dead-but-wrong storage into dead-but-right storage. It moves no number now.

The formula turned out to be hand-written at **twelve** sites, seven correct and five not; all twelve now call one helper.

**Part 2 of this issue is untouched** and stays open behind #1388: the log-domain site whose distance *gradient* is annihilated in float32. It cannot be fixed by rescaling — lifting the cotangent above 1.1755e-38 needs `L_nu > 4.2e38` against a float32 max of 3.4e38 — so the derivative has to be taken in log space, not merely the value.
